### ISSUE TITLE
[DeepEP] shared_experts cannot overlap with deepep.combine()

### ISSUE THREAD

**Issue by goldhuang** [pytorch/torchtitan#2298]
Profiling results - [ep-rank3-deepep.json](https://github.com/user-attachments/files/24949436/ep-rank3-deepep.json)

### Bug description

[run_deepep.py](https://github.com/user-attachments/files/24949358/run_deepep.py)

Hi team, I find that deepep communication (dispatch & combine) time is totally wasted in the current deepep integration. I think at least we should make shared_experts either overlap with dispatch or combine during forward.

To reproduce, run the attached file with 
```
torchrun --nproc_per_node=8 run_deepep.py
```
(All the code in the file is from latest torchtitan. Sorry I cannot directly import torchtitan package due to dependency issue.)

### Versions

I'm using PyTorch 2.9.1. Let me know if you cannot reproduce the issue when running with PyTorch nightly.

**Comment by shuhuayu** [comment]
@goldhuang Thanks for pointing this out. We are still considering whether we should add this feature into deepep integration, and indeed in our offline tests overlap communication with `shared_experts` computation can improve mfu by around 2%.  To unblock youself, you may need to customize a little bit for now.

**Comment by elfiegg** [comment]
Heads up that @vivekgoe is working in progress to overlap shared experts with communications on Torchtitan.

**Comment by vivekgoe** [comment]
I will push a PR with the change to overlap combine with shared_experts during forward. @shuhuayu 2% MFU integration was seen in E2E full model tests or unit-tests? Duration of shared_experts compute is not large and for full model it may not give large gain.

**Comment by shuhuayu** [comment]
> I will push a PR with the change to overlap combine with shared_experts during forward. [@shuhuayu](https://github.com/shuhuayu) 2% MFU integration was seen in E2E full model tests or unit-tests? Duration of shared_experts compute is not large and for full model it may not give large gain.

E2E full model tests in a 16b model setup. The gain depends on the amount of shared_experts and may vary in your tests.

**Comment by goldhuang** [comment]
@vivekgoe I think it's better to have the overlap support for both dispatch and combine. E.g. DeepSeek 16B model uses 2 shared experts and 6 activate experts. We can use one to overlap dispatch and the other to overlap combine.
Or, provide the option to either overlap dispatch or combine. Thanks!

**Comment by tianyu-l** [comment]
We used to plan to use DualPipe to overlap EP comms with computation. If that's no longer the plan, we can consider shared experts overlapping. cc @soulitzer @H-Huang

**Comment by rakkit** [comment]
> We used to plan to use DualPipe to overlap EP comms with computation.

I am curious at what scale shared experts can completely overlap the EP.  I read at somewhere times ago for "big" models, those EP runs on NVLINK + IB/RDMA, it might be difficult to overlap purely based on shared experts. (unless with a giant big shared experts) if that is the case we may still need PP

**Comment by vivekgoe** [comment]
@goldhuang Because of the way DeepEP handles streams it is relatively straightforward and low-risk to overlap combine and shared_experts whereas overlap between shared_experts and dispatch will need additional stream creation from Torchtitan and  maybe changes to DeepEP code as well.  In my opinion combine should always be able to hide shared_expert compute completely, did a quick check for DeepSeek 16B (see snapshot below),

<img width="1675" height="732" alt="Image" src="https://github.com/user-attachments/assets/014b000b-c237-46c3-92c0-b672f3fc9088" />

**Comment by goldhuang** [comment]
@vivekgoe I came up with the same solution for overlapping combine in fwd.
I'm working on overlapping combine in both fwd and bwd now. I'm passing shared_expert module to combine, and do combine- shared_experts-sync in both combine fwd and bwd. Basically fuse deepep combine with shared experts. To do it, seems I have to stop wrapping combine as a custom operator but using a autograd funciton instead. I think we can also follow this pattern to overlap with dispatch fwd+bwd.

**Comment by elfiegg** [comment]
@goldhuang I vaguely remember you mentioned you’d probed memory consumption and concluded that including DeepEP as part of SAC doesn’t really make sense. Could you share a bit more detail on that?

Registering a custom op was mainly aimed at supporting SAC, but it’s introduced a few issues - e.g. interface mismatches between DeepEP and custom ops, and the stream-overlap concerns you pointed out. If we have a strong case that DeepEP isn’t actually needed in SAC in practice, we could switch to an autograd function instead.

I also have another PR in progress to support HybridEP, so I’d like to align the design with whatever direction we decide

**Comment by goldhuang** [comment]
@elfiegg 

1. If we could overlap computation with communication during fwd, we don't need SAC.
2. Finally, training efficiency is about throughput. If we use a lot more memory, meaning we use much smaller batch size, throughput could actually drop although iteration time is less. In that case, it does not make sense to use more memory if it finally lead to worse throughput.

I'm working a MoE DiT model. At least in my case, SAC does not make sense. And as I mentioned in item 1, computation-communication overlapping is always most important.

**Comment by shuhuayu** [comment]
IIRC the previous discussion on SAC is to avoid SAC completely to save memory for things like batch size, not in particular whether deep ops should be included in SAC save list. The latter, which I experimented on 16b models in https://github.com/pytorch/torchtitan/pull/2296, does not make big difference for small models. For large models, we don't know yet. I think it still makes sense to keep it for user choice, dispatch and combine are communication heavy and overlapping with shared_experts is probably not hiding it. 

@elfiegg  I think @soulitzer made a point in your hybridep pr https://github.com/pytorch/torchtitan/pull/2207, there is a new feature that may allow use to autograd instead of ops (to remove global cache).

**Comment by elfiegg** [comment]
@shuhuayu I noticed that experimental feature, it overall makes me wonder the necessarily of wrapping DeepEP in a custom op and add it to SAC list. Here is the detail for DeepseekV3 671B model:

```
# dispatch returns 5 tensors:
recv_x:        [num_recv_tokens, hidden_dim]  # Main activation
recv_indices:  [num_recv_tokens, top_k]       # Small integer tensor
recv_scores:   [num_recv_tokens, top_k]       # Small float tensor  
num_recv_per_expert: [num_local_experts]      # Tiny tensor
cache_id:      [1]                            # Single CPU int64
```
combine returns
```
# combine returns:
combined_x: [local_batch_size × seq_len, hidden_dim] 
```

For DeepSeek-V3 configuration:
```
hidden_dim = 7168
seq_len = 4096  
top_k = 8

# Tokens received per rank (roughly uniform distribution):
num_recv_tokens ≈ local_batch_size  * seq_len * top_k 
```
So let's say local_batch_size = 4, we would need:

Dispatch Memory (5 tensors saved by SAC)
| Tensor | Shape | Size |
|--------|-------|------|
| recv_x | [131072, 7168] | 1.75 GiB |
| recv_indices | [131072, 8] | 4 MiB |
| recv_scores | [131072, 8] | 4 MiB |
| num_recv_per_expert | [4] | |
| cache_id | [1] | |
| **Dispatch Total** | | **1.76 GiB** |

Combine Memory (1 tensor saved by SAC)
| Tensor | Shape | Size |
|--------|-------|------|
| combined_x | [16384, 7168] | 224 MiB |
| **Combine Total** | | **0.22 GiB** |

So we have in total ~2GB consumed per one layer. Does this make sense to you? Question is - how does activation checkpointing actually work. How many layers will be saved at once?

**Comment by goldhuang** [comment]
@elfiegg You are using 2GB * num_blocks extra memory by doing SAC.

**Comment by elfiegg** [comment]
@goldhuang Just double checking my understanding: does `num_blocks` mean the number of MoE layers? I assume we're holding memory for all MoE blocks until the last layer's backward pass triggers the release. Is that the case in torchtitan, or are there other memory saving optimizations at play here?

**Comment by elfiegg** [comment]
Let's say all forward layers need to be saved until the last layer starts to release in the backward process - that'd be at peak ~2GB * 58 layers = ~116G per device, which is almost certain not going to be practical.

It may make some sense for very small model, where usually the model isn't computation bound but CPU/communication bound (Increasing batch size will help address this bottleneck though again limit how much we SAC?).

**Comment by shuhuayu** [comment]
@elfiegg @goldhuang  Thanks for the reasoning! The activations from `dispatch` and `combine` are indeed very big as shown above due to the `top_k` expansion, and I think pipeline parallel is needed to share the memory load if these activations are saved (like if we have pipeline degree=4, each rank will have ~116/4=29GB). My understanding is that our discussion is not specific to deepep but also applies to the standard all-to-all expert parallel communication backend, and the decision on whether to save dispatch/combine really matters.

I am leaning toward leaving this fine-grained recomputation control to users. If we implement dispatch/combine as custom ops, then when SAC is enabled with the op option, users can easily choose whether to save or recompute these activations based on their regime, e.g., communication-bound, memory-bound, and whether all-to-all communication is fully overlapped.

Some references that might be useful: 
1. Megatron has this `moe_act` option that allows users to recompute the mlp functions only in moe https://swift.readthedocs.io/en/v3.7/Instruction/Megatron-SWIFT-Training.html. 
2. A more recent report on moe model training (https://arxiv.org/abs/2505.11432). The communication in MoE is redone during backward but the all2all is replaced by all gather + reduce scatter in some regimes, which is shown to be more efficient in ring topology (Figure 7).

**Comment by soulitzer** [comment]
>  there is a new feature that may allow use to autograd instead of ops (to remove global cache).

Just to clarify, although this removes the need for a global cache, you still need a custom op for SAC to work

**Comment by elfiegg** [comment]
@shuhuayu From past experiments - For PP=4, EP=32, local BS = 16, GA=1, seq_len=4096, with full activation checkpointing, the first step peak memory consumption is already around 67.58GiB(85.42%). I get the idea but my point is this doesn't seem to be practical to SAC DeepEP ops.  
I'd love to understand how wrapping DeepEP in a custom op prevents us from doing comm + comp overlapping though, which might do us a 1-2% MFU gain end to end.

**Comment by shuhuayu** [comment]
>For PP=4, EP=32, local BS = 16, GA=1, seq_len=4096, with full activation checkpointing, the first step peak memory consumption is already around 67.58GiB(85.42%). I get the idea but my point is this doesn't seem to be practical to SAC DeepEP ops.

@elfiegg That's a good point. I am leaning towards leaving this option to users (whether saving the outputs of deepep ops or standard all-to-all communication) since I don't have a strong case to remove this option. Like, maybe there are users training in multi-node setups where all-to-all communication is a bottleneck. On the other hand, do you think keeping this option will block us on other features? Would like to know your thoughts on this.  

>how wrapping DeepEP in a custom op prevents us from doing comm + comp overlapping

I don't see it how it prevents us doing so. I am open to add comm + comp overlapping given its benefits.

**Comment by elfiegg** [comment]
@shuhuayu Sounds good, I did a bit of research and I think I might have misunderstood @goldhuang 's point where he said he needs to stop wrapping combine as a custom operator but using an autograd funciton instead for overlapping shared experts in the backwards

It looks like if we want to overlap backward SharedExpert with communications, we need to wrap the entire MoE in an autograd function. Because the order in which autograd calls the different backward functions across independent ops is something we can't control. Autograd uses a LIFO queue and we can't tell it "process shared_expert backward before expert backward."

So, it's less about a design preference between DeepEP custom ops and autograd functions, and more about a functional requirement to manage the MoE execution flow

**Comment by goldhuang** [comment]
@elfiegg Yes, we need to embed shared expert's fwd and bwd into deepep's autograd func. (You cannot embed it to a custom op because of definition limitation of a custom op.)
I'm able to overlap shared expert fwd with combine fwd and shared expert bwd with combine bwd by embedding shared expert into combine autograd function. My implementation passed accuracy, and memory profiling, also confirmed overlap by torch profiling.
I did some hacks to work around activation checkpointing to call a nn.module inside an autograd fwd and bwd, without extra memory or computation. Looking forward to the official implementation from your team, if there is a plan.
cc: @shuhuayu

**Comment by soulitzer** [comment]
> Autograd uses a LIFO queue and we can't tell it "process shared_expert backward before expert backward."

We actually do have an API that can tell autograd to change the priority of nodes during bw compute

https://github.com/pytorch/pytorch/blob/8e65dfa6f2c451dc3c71d76d68796d1271c72772/test/test_autograd.py#L15431

Today it is a low-level interface, and not public facing, but curious if could avoid the need for this mega autograd Function / custom op.
